hedera-node/docs/design/services/token-service/airdrops/airdrop-pending-state.md

```diff
@@ -0,0 +1,123 @@
+
+Details for all of them would be covered in their own design docs, but here is a general overview of how they would interact with the pending state:
+- `TokenAirdrop` would add a new airdrop to the pending state
+- `TokenClaimAirdrop` would have a handle logic, which would take and remove a pending airdrop from this state, create a synthetic `CryptoTransfer` and execute it
+- `TokenCancelAirdrop` operation would remove an existing airdrop from the state
+
+### Structure
+
+We should define the structure of pending changes that would be kept in the `InitialModServiceTokenSchema` state, so we would reuse the token’s serialized state and add pending airdrops records in it. We could use the type defined in the protobuf records for the airdrop information itself. We should use the protobuf types since they are used for state serialization. So, the structure could look like this:
+
+`StateDefinition.*onDisk*(*AIRDROPS_KEY*, PendingAirdropId.PROTOBUF, PendingAirdropValue.PROTOBUF, MAX__AIRDROPS);`
+
+We can have a case, where we have multiple airdrops for the same fungible token. In this case the  `PendingAirdropValue` would be updated with the new aggregated value.
+
+### Stores
+
+We need to define a special `WritableStore` for the airdrops, so that we can manipulate the content in them. This would be the `WritableAirdropStore`.
+
+It would extend `ReadableAirdropStoreImpl` that implements the following interface:
+
+```java
+public interface ReadableAirdropStore {
+	PendingAirdropValue getFungibleAirdropAmount(@NonNull final PendingAirdropId airdropId);
+	long sizeOfState();
+	void warm(@NonNull final PendingAirdropId airdropId);
+    List<PendingAirdropId> getPendingAirdropIdsBySender(final AccountId sender);
+}
+```
+
+The `ReadableAirdropStoreImpl` would keep an instance of `ReadableKVState<PendingAirdropId, PendingAirdropValue> airdropsState`.
+
+An additional method `getPendingAirdropIdsBySender` would be added to the `ReadableAirdropStore` to get all `PendingAirdropIds` for a specific sender. This would be done via filtering all `PendingAirdropIds` kept in state by the `sender` that is passed as a parameter.
+
+This would be useful for traversing all sender's airdrops when their account expires or gets deleted. Then we should iterate over all of their pending airdrops and cancel them.
+
+The `WritableAirdropStore` would add additional methods for operating over the airdrops state, so it would have the following methods:
+
+```java
+public class WritableAirdropStore {
+
+    void put(@NonNull final PendingAirdropId airdropId, @NonNull final PendingAirdropValue airdropValue) {
+        // invoked by the `TokenAirdropTransaction`
+    }
+
+    void remove(@NonNull PendingAirdropId airdropId) {
+        // invoked by the `TokenCancelAirdrop`
+    }
+}
+```
+
+### Operations
+
+1. Fetching airdrop values from the collections. This would be applicable only for fungible airdrops (for both `TokenAirdrop` and `TokenClaimAirdrop` transactions).
```

Comment:
Does this mean we cannot determine if there is a pending airdrop for an NFT?

Reply:
No, this method will return the TokenAirdropValue is applicable only for fungible airdrops. For NFT airdrops we don't have such a value. It just keeps the amount of fungible tokens we want to airdrop.

I added a clarification sentence, but basically for an NFT airdrop we would have all of the needed information in the PendingAirdropId itself and won't need to fetch something additional.